modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/persistence/checkpoint/CheckpointReadWriteLock.java

```diff
@@ -182,14 +199,35 @@ public boolean hasQueuedWriters() {
     }
 
     private void onReadLock(long start, boolean taken) {
-        long elapsed = coarseCurrentTimeMillis() - start;
+        readLockWaitingThreadsGauge.decrementAndGet();
+
+        long coarseCurrentTimeMillis = coarseCurrentTimeMillis();
+        long elapsed = coarseCurrentTimeMillis - start;
 
         if (taken) {
+            if (checkpointReadLockAcquiredTime.get() == null) {
```

Comment:
Please add a comment, it took me a while to figure out why we need this check. Or maybe we should base this code on the value of `checkpointReadLockHoldCount` that we read in the next statement instead, this will make code simpler